src/integrations/contact-loaders/ngpvan/index.js

```diff
@@ -0,0 +1,381 @@
+import { finalizeContactLoad } from "../helpers";
+import { getConfig } from "../../../server/api/lib/config";
+import { parseCSVAsync } from "../../../workers/parse_csv";
+import { failedContactLoad } from "../../../workers/jobs";
+import HttpRequest from "../../../server/lib/http-request.js";
+import Van from "./util";
+
+export const name = "ngpvan";
+
+export const DEFAULT_NGP_VAN_MAXIMUM_LIST_SIZE = 75000;
+export const DEFAULT_NGP_VAN_CACHE_TTL = 300;
+export const DEFAULT_NGP_VAN_EXPORT_JOB_TYPE_ID = 8;
+export const DEFAULT_PHONE_NUMBER_COUNTRY = "US";
+
+export function displayName() {
+  return "NGP VAN";
+}
+
+export function serverAdministratorInstructions() {
+  return {
+    description:
+      "Load contacts from VAN (and LAN, everyaction, etc.) saved lists",
+    setupInstructions:
+      "Get an APP name and API key for your VAN account. Add them to your config, along with NGP_VAN_WEBHOOK_BASE_URL. In most cases the defaults for the other environment variables will work",
+    environmentVariables: [
+      "NGP_VAN_API_KEY",
+      "NGP_VAN_API_BASE_URL",
+      "NGP_VAN_APP_NAME",
+      "NGP_VAN_CACHE_TTL",
+      "NGP_VAN_EXPORT_JOB_TYPE_ID",
+      "NGP_VAN_MAXIMUM_LIST_SIZE",
+      "NGP_VAN_WEBHOOK_BASE_URL"
+    ]
+  };
+}
+
+export const handleFailedContactLoad = async (
+  job,
+  ingestDataReference,
+  message
+) => {
+  // eslint-disable-next-line no-console
+  console.error(message);
+  await failedContactLoad(job, null, JSON.stringify(ingestDataReference), {
+    errors: [message],
+    ...ingestDataReference
+  });
+};
+
+export async function available(organization, user) {
+  // / return an object with two keys: result: true/false
+  // / these keys indicate if the ingest-contact-loader is usable
+  // / Sometimes credentials need to be setup, etc.
+  // / A second key expiresSeconds: should be how often this needs to be checked
+  // / If this is instantaneous, you can have it be 0 (i.e. always), but if it takes time
+  // / to e.g. verify credentials or test server availability,
+  // / then it's better to allow the result to be cached
+
+  const result =
+    !!getConfig("NGP_VAN_API_KEY", organization) &&
+    !!getConfig("NGP_VAN_APP_NAME", organization) &&
+    !!getConfig("NGP_VAN_WEBHOOK_BASE_URL", organization);
+
+  if (!result) {
+    console.log(
+      "ngpvan contact loader unavailable. Missing one or more required environment variables."
+    );
+  }
+
+  return {
+    result,
+    expiresSeconds: 60
+  };
+}
+
+export function addServerEndpoints(expressApp) {
+  // this method allows us to add server endpoint to serve as a webhook for external systems
+  // For example, VAN will call this webhook when a contact export is complete on their end
+  // Currently, this is implemented by VAN as a placeholder, it carries no meaningful data and
+  // is included here for illustrative purposes.
+  expressApp.post(
+    "/ingest-data/ngpvan/:jobId/:maxContacts/:vanListId",
```

Comment:
this should start `/integration/ngpvan/` instead of `/ingest-data/ngpvan` or is this a hard-coded endpoint path from VAN?
See https://github.com/MoveOnOrg/Spoke/blob/main/docs/HOWTO-use-contact-loaders.md#using-addserverendpoints
see components/Login.jsx for why -- basically we reserve some prefixes so that the general spoke app knows where to stay clear of endpoints for integrations but can also add new ones without worrying about clobbering an existing endpoint.

The Login support will also make the kind of experience of clicking from NGPVAN to launch the Spoke app, even through a login, work successfully.

Reply:
dded1bdbe164485db44920abdb49968d5a668dc7